## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [4]:
%pip install pandas
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

Looking in indexes: https://vrajagopalan:****@artifactory-uw2.adobeitc.com/artifactory/api/pypi/pypi-wf-dev/simple

[notice] A new release of pip is available: 24.0 -> 24.1.1
[notice] To update, run: python3.11 -m pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [7]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [9]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [16]:
# Use this space to do any additional research
# and familiarize yourself with the data.
copy_df = df.copy()
copy_df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [34]:
# What three item categories had the most entries?
top_item_categories = copy_df['category'].value_counts().head(3)
top_item_categories

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [37]:
# For the category with the most entries, which subcategory had the most entries?
category_most_entries = top_item_categories.idxmax()
top_sub_category = copy_df[copy_df['category'] == category_most_entries]['subcategory'].value_counts().idxmax()
top_sub_category

'bathroom supplies'

In [67]:
# Which five clients had the most entries in the data?
clients_with_most_entries = copy_df['client_id'].value_counts().head(5)
clients_with_most_entries

client_id
33615    220
66037    211
46820    209
24741    207
38378    207
Name: count, dtype: int64

In [97]:
# Store the client ids of those top 5 clients in a list.
top_5_clients = copy_df['client_id'].value_counts().head(5).index.tolist()
top_5_clients

[33615, 66037, 46820, 24741, 38378]

In [72]:
# How many total units (the qty column) did the client with the most entries order order?
clients_with_most_entries = copy_df['client_id'].value_counts().idxmax()
total_units_client_most_entries = (copy_df[copy_df['client_id'] == clients_with_most_entries]['qty']).sum()
total_units_client_most_entries

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [78]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
import pandas as pd
new_df = pd.read_csv('Resources/client_dataset.csv')
new_df['sub_total'] = new_df['unit_price'] * new_df['qty']
new_df.head()




first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  sub_total  
0     1096.80     762.71         7.50  105            1  115164.00  
1       24.95      15.09         1.49   21            0     523.95  
2       13.52       7.86         1.68   39            6     527.28  
3       36.42      24.85         1.23   29            3    1056.18  
4      195.10     108.17        46.43   20            1    3902.00

In [79]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.

# create a function to determine the shipping price
def shipping_price(weight):
    if weight > 50:
        return weight * 7
    else:
        return weight * 10
    
new_df['shipping_price'] =   new_df['unit_weight'].apply(shipping_price)
new_df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  sub_total  \
0     1096.80     762.71         7.50  105            1  115164.00   
1       24.95      15.09         1.49   21            0     523.95   
2       13.52       7.86         1.68   39            6     527.28   
3       36.42      24.85         1.23   29            3    1056.18   
4      195.10     108.17        46.43   20            1    3902.00   

   shipping_price  
0            75.0  
1            14.9  
2            16.8  
3            12.3  
4           464.3

In [81]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%

new_df['total_price'] = (new_df['sub_total']+ new_df['shipping_price']) * 1.0925
new_df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number  sub_total  shipping_price  \
0     762.71         7.50  105            1  115164.00            75.0   
1      15.09         1.49   21            0     523.95            14.9   
2       7.86         1.68   39            6     527.28            16.8   
3      24.85         1.23   29            3    1056.18            12.3   
4     108.17        46.43   20            1    3902.00           464.3   

     total_price  
0  125898.607500  
1     588.693625  
2     594.407400  
3    1167.314400  
4    4770.182750  

[5 rows x 21 columns]

In [83]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
new_df['line_cost'] = new_df['unit_cost'] * new_df['qty'] + new_df['shipping_price']
new_df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number  sub_total  shipping_price    total_price  \
0         7.50  105            1  115164.00            75.0  125898.607500   
1         1.49   21            0     523.95            14.9     588.693625   
2         1.68   39            6     527.28            16.8     594.407400   
3         1.23   29            3    1056.18            12.3    1167.314400   
4        46.43   20            1    3902.00           464.3    4770.182750   

   line_cost  
0   80159.55  
1     331.79  
2     323.34  
3     732.95  
4    2627.70  

[5 rows x 22 columns]

In [85]:
# Create a column for the profit of each line using line cost and line price

new_df['line_profit'] = new_df['total_price'] - new_df['line_cost']
new_df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number  sub_total  shipping_price    total_price  line_cost  \
0            1  115164.00            75.0  125898.607500   80159.55   
1            0     523.95            14.9     588.693625     331.79   
2            6     527.28            16.8     594.407400     323.34   
3            3    1056.18            12.3    1167.314400     732.95   
4            1    3902.00           464.3    4770.182750    2627.70   

    line_profit  
0  45739.057500  
1    256.903625  
2    271.067400  
3    434.364400  
4   2142.482750  

[5 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [95]:
# Check your work using the totals above
#first_order_verification = new_df[new_df['order_id'] == 2742071]['total_price'].sum()

#first_order_verification

order_2742071 = new_df[new_df['order_id'] == 2742071]
order_2173913 = new_df[new_df['order_id'] == 2173913]
order_6128929 = new_df[new_df['order_id'] == 6128929]
# order_2742071
# create a function to determine the shipping price
def ship_price(weight):
    if weight > 50:
        return weight * 7
    else:
        return weight * 10
    
def cal_total_price(row):
    subtotal = row['unit_price'] * row['qty']
    shipping_price = ship_price(row['unit_weight'])
    total_price = (subtotal + shipping_price) * 1.0925
    return total_price

order_2742071['new_total_price'] = order_2742071.apply(cal_total_price, axis=1)
order_2742071
order_2173913['new_total_price'] = order_2173913.apply(cal_total_price, axis=1)
order_2173913
order_6128929['new_total_price'] = order_6128929.apply(cal_total_price, axis=1)
order_6128929

/var/folders/q1/tmt8zmsx0l30ykyt0bz1q_gm0000gr/T/ipykernel_37985/1806481157.py:23: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  order_2742071['new_total_price'] = order_2742071.apply(cal_total_price, axis=1)
/var/folders/q1/tmt8zmsx0l30ykyt0bz1q_gm0000gr/T/ipykernel_37985/1806481157.py:25: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  order_2173913['new_total_price'] = order_2173913.apply(cal_total_price, axis=1)
/var/folders/q1/tmt8zmsx0l30ykyt0bz1q_gm0000gr/T/ipykernel_37985/1806481157.py:27: SettingWith

first  last                       job                   phone  \
3422   Dorothy  Hall  Journalist, broadcasting  001-338-609-8495x03504   
13672  Dorothy  Hall  Journalist, broadcasting  001-338-609-8495x03504   
14517  Dorothy  Hall  Journalist, broadcasting  001-338-609-8495x03504   
16295  Dorothy  Hall  Journalist, broadcasting  001-338-609-8495x03504   
21141  Dorothy  Hall  Journalist, broadcasting  001-338-609-8495x03504   
37265  Dorothy  Hall  Journalist, broadcasting  001-338-609-8495x03504   
37949  Dorothy  Hall  Journalist, broadcasting  001-338-609-8495x03504   
38542  Dorothy  Hall  Journalist, broadcasting  001-338-609-8495x03504   
48834  Dorothy  Hall  Journalist, broadcasting  001-338-609-8495x03504   
49353  Dorothy  Hall  Journalist, broadcasting  001-338-609-8495x03504   

                                        email  client_id  order_id  \
3422   dorothy.per.1343@mitchell-chambers.com      88257   6128929   
13672  dorothy.per.1343@mitchell-chambers.com      88257   6128929   
14517  dorothy.per.1343@mitchell-chambers.com      88257   6128929   
16295  dorothy.per.1343@mitchell-chambers.com      88257   6128929   
21141  dorothy.per.1343@mitchell-chambers.com      88257   6128929   
37265  dorothy.per.1343@mitchell-chambers.com      88257   6128929   
37949  dorothy.per.1343@mitchell-chambers.com      88257   6128929   
38542  dorothy.per.1343@mitchell-chambers.com      88257   6128929   
48834  dorothy.per.1343@mitchell-chambers.com      88257   6128929   
49353  dorothy.per.1343@mitchell-chambers.com      88257   6128929   

       order_date  order_week  order_year  ... unit_cost unit_weight   qty  \
3422   2023-03-12          10        2023  ...     11.75        7.61   671   
13672  2023-03-12          10        2023  ...     90.12        7.05    40   
14517  2023-03-12          10        2023  ...     98.35        6.28  1089   
16295  2023-03-12          10        2023  ...     42.72        0.78    44   
21141  2023-03-12          10        2023  ...     23.64        4.60    26   
37265  2023-03-12          10        2023  ...      1.14        6.00    45   
37949  2023-03-12          10        2023  ...    470.05        6.55   344   
38542  2023-03-12          10        2023  ...      5.50        1.38     8   
48834  2023-03-12          10        2023  ...     98.70        0.55   664   
49353  2023-03-12          10        2023  ...     72.54        4.26  1663   

       line_number  sub_total  shipping_price    total_price  line_cost  \
3422             3    8890.75            76.1    9796.283625    7960.35   
13672            6    5459.60            70.5    6041.634250    3675.30   
14517            9  117187.29            62.8  128095.723325  107165.95   
16295            8    3439.92             7.8    3766.634100    1887.48   
21141            2     731.12            46.0     849.003600     660.64   
37265            1      73.35            60.0     145.684875     111.30   
37949            0  320470.40            65.5  350185.470750  161762.70   
38542            4      48.72            13.8      68.303100      57.80   
48834            5   90974.64             5.5   99395.802950   65542.30   
49353            7  141288.48            42.6  154404.204900  120676.62   

         line_profit  new_total_price  
3422     1835.933625      9796.283625  
13672    2366.334250      6041.634250  
14517   20929.773325    128095.723325  
16295    1879.154100      3766.634100  
21141     188.363600       849.003600  
37265      34.384875       145.684875  
37949  188422.770750    350185.470750  
38542      10.503100        68.303100  
48834   33853.502950     99395.802950  
49353   33727.584900    154404.204900  

[10 rows x 24 columns]

In [ ]:
# Check your work using the totals above
#first_order_verification = new_df[new_df['order_id'] == 2742071]['total_price'].sum()

#first_order_verification

order_2742071 = new_df[new_df['order_id'] == 2742071]
order_2173913 = new_df[new_df['order_id'] == 2173913]
order_6128929 = new_df[new_df['order_id'] == 6128929]
# order_2742071
# create a function to determine the shipping price
def ship_price(weight):
    if weight > 50:
        return weight * 7
    else:
        return weight * 10
    
def cal_total_price(row):
    subtotal = row['unit_price'] * row['qty']
    shipping_price = ship_price(row['unit_weight'])
    total_price = (subtotal + shipping_price) * 1.0925
    return total_price

order_2742071['new_total_price'] = order_2742071.apply(cal_total_price, axis=1)
order_2742071
order_2173913['new_total_price'] = order_2742071.apply(cal_total_price, axis=1)
order_2173913
order_6128929['new_total_price'] = order_2742071.apply(cal_total_price, axis=1)
order_2173913

/var/folders/q1/tmt8zmsx0l30ykyt0bz1q_gm0000gr/T/ipykernel_37985/1871383533.py:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  order_2742071['new_total_price'] = order_2742071.apply(cal_total_price, axis=1)


first   last                                       job  \
3446   Kimberly  Blair  Geographical information systems officer   
6287   Kimberly  Blair  Geographical information systems officer   
20623  Kimberly  Blair  Geographical information systems officer   
23840  Kimberly  Blair  Geographical information systems officer   
24671  Kimberly  Blair  Geographical information systems officer   
42470  Kimberly  Blair  Geographical information systems officer   

                    phone                           email  client_id  \
3446   477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
6287   477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
20623  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
23840  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
24671  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
42470  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   

       order_id  order_date  order_week  order_year  ... unit_cost  \
3446    2742071  2023-02-27           9        2023  ...      3.05   
6287    2742071  2023-02-27           9        2023  ...    407.68   
20623   2742071  2023-02-27           9        2023  ...     80.78   
23840   2742071  2023-02-27           9        2023  ...     35.41   
24671   2742071  2023-02-27           9        2023  ...     54.86   
42470   2742071  2023-02-27           9        2023  ...     61.07   

      unit_weight  qty  line_number  sub_total  shipping_price    total_price  \
3446        10.46   77            5     430.43           104.6     584.520275   
6287         2.60  260            0  119652.00            26.0  130748.215000   
20623        2.27   12            3    1621.44            22.7    1796.222950   
23840        0.54   22            1    1046.98             5.4    1149.725150   
24671        1.57   24            2    1432.56            15.7    1582.224050   
42470        3.04   44            4    3615.92            30.4    3983.604600   

       line_cost   line_profit  new_total_price  
3446      339.45    245.070275       584.520275  
6287   106022.80  24725.415000    130748.215000  
20623     992.06    804.162950      1796.222950  
23840     784.42    365.305150      1149.725150  
24671    1332.34    249.884050      1582.224050  
42470    2717.48   1266.124600      3983.604600  

[6 rows x 24 columns]

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [103]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
top_clients_spending = new_df[new_df['client_id'].isin(top_5_clients)].groupby('client_id')['total_price'].sum()
# Convert the result to a DataFrame for better display
top_clients_spending_df = top_clients_spending.reset_index().rename(columns={'client_id': 'Client ID', 'total_price': 'Total Spending'})

# Format the DataFrame with commas
top_clients_spending_df['Total Spending'] = top_clients_spending_df['Total Spending'].apply(lambda x: f"${x:,.2f}")
top_clients_spending_df

Client ID  Total Spending
0      24741  $76,678,479.05
1      33615   $6,392,918.35
2      38378   $9,172,866.14
3      46820   $8,006,332.53
4      66037   $8,746,255.31

In [106]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
top_clients_df = new_df[new_df['client_id'].isin(top_5_clients)]
# Create the summary DataFrame
summary_df = top_clients_df.groupby('client_id').agg({
    'qty': 'sum',
    'shipping_price': 'sum',
    'total_price': 'sum',
    'line_profit': 'sum'
}).reset_index()

# Rename columns for clarity
summary_df.columns = ['Client ID', 'Total Units Purchased', 'Total Shipping Price', 'Total Revenue', 'Total Profit']

# Format the DataFrame for better readability
summary_df['Total Revenue'] = summary_df['Total Revenue'].apply(lambda x: f"${x:,.2f}")
summary_df['Total Shipping Price'] = summary_df['Total Shipping Price'].apply(lambda x: f"${x:,.2f}")
summary_df['Total Profit'] = summary_df['Total Profit'].apply(lambda x: f"${x:,.2f}")
summary_df

Client ID  Total Units Purchased Total Shipping Price   Total Revenue  \
0      24741                 239862            $9,365.60  $76,678,479.05   
1      33615                  64313           $12,609.40   $6,392,918.35   
2      38378                  73667           $11,895.00   $9,172,866.14   
3      46820                  75768           $11,094.80   $8,006,332.53   
4      66037                  43018           $10,017.30   $8,746,255.31   

     Total Profit  
0  $36,106,662.11  
1   $2,033,979.93  
2   $2,955,705.56  
3   $2,589,494.93  
4   $3,126,906.88

In [107]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 


# Define a function that converts a dollar amount to millions.


# Apply the currency_format_millions function to only the money columns. 


# Rename the columns to reflect the change in the money format. 

# Filter data for the top 5 clients
top_clients_data = new_df[new_df['client_id'].isin(top_5_clients)]

# Create the summary DataFrame
summary_df = top_clients_data.groupby('client_id').agg({
    'qty': 'sum',
    'shipping_price': 'sum',
    'total_price': 'sum',
    'line_profit': 'sum'
}).reset_index()

# Rename columns for clarity
summary_df.columns = ['Client ID', 'Total Units Purchased', 'Total Shipping Price', 'Total Revenue', 'Total Profit']

# Define the money columns
money_columns = ['Total Shipping Price', 'Total Revenue', 'Total Profit']

# Define a function that converts a dollar amount to millions
def to_millions(amount):
    return amount / 1_000_000

# Apply the function to the money columns
for column in money_columns:
    summary_df[column] = summary_df[column].apply(to_millions)


summary_df['Total Revenue'] = summary_df['Total Revenue'].apply(lambda x: f"${x:,.2f}")
summary_df['Total Shipping Price'] = summary_df['Total Shipping Price'].apply(lambda x: f"${x:,.2f}")
summary_df['Total Profit'] = summary_df['Total Profit'].apply(lambda x: f"${x:,.2f}")
# Rename the columns to reflect the change in the money format
summary_df = summary_df.rename(columns={
    'Total Shipping Price': 'Total Shipping Price (Millions)',
    'Total Revenue': 'Total Revenue (Millions)',
    'Total Profit': 'Total Profit (Millions)'
})
summary_df

Client ID  Total Units Purchased Total Shipping Price (Millions)  \
0      24741                 239862                           $0.01   
1      33615                  64313                           $0.01   
2      38378                  73667                           $0.01   
3      46820                  75768                           $0.01   
4      66037                  43018                           $0.01   

  Total Revenue (Millions) Total Profit (Millions)  
0                   $76.68                  $36.11  
1                    $6.39                   $2.03  
2                    $9.17                   $2.96  
3                    $8.01                   $2.59  
4                    $8.75                   $3.13

In [109]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_summary_df = summary_df.sort_values(by='Total Profit (Millions)', ascending=False)
sorted_summary_df

Client ID  Total Units Purchased Total Shipping Price (Millions)  \
0      24741                 239862                           $0.01   
4      66037                  43018                           $0.01   
2      38378                  73667                           $0.01   
3      46820                  75768                           $0.01   
1      33615                  64313                           $0.01   

  Total Revenue (Millions) Total Profit (Millions)  
0                   $76.68                  $36.11  
4                    $8.75                   $3.13  
2                    $9.17                   $2.96  
3                    $8.01                   $2.59  
1                    $6.39                   $2.03

In [ ]:
# type: ignore
The analysis of the top 5 clients reveals the following insights: 

Client ID 24741 is the highest spender with a total revenue of $76.68 million and the highest total profit of $36.11 million.
Shipping costs vary slightly among the clients, with the highest shipping cost incurred by Client ID 33615 at $12,609.
The total units purchased by each client show significant variability, highlighting different order sizes and purchasing behaviors.
These findings provide valuable insights into client spending patterns, profitability, and shipping costs, which can help tailor business strategies for high value clients